### PR TITLE
add debugUri command option

### PR DESCRIPTION
**Pull request by zhfnjust**

### CODE REVIEW

File: src/compilerWrapper.ts [hunk 1]
```diff
@@ -9,19 +9,18 @@ import { path2uri } from './utils';
...
+const INTERNAL_ERR_REG =  /Internal error:(?<message>.+)/;
+
+
 //SOURCE_REG parser src eg: [0:6:3:8:4#Bar.constructor:0]
 const SOURCE_REG =  /^(?<fileIndex>-?\d+):(?<line>\d+):(?<col>\d+):(?<endLine>\d+):(?<endCol>\d+)(#(?<tagStr>.+))?/;
-const INTERNAL_ERR_REG =  /Internal error:(?<message>.+)/;
 
 // see VERSIONLOG.md
-const CURRENT_CONTRACT_DESCRIPTION_VERSION = 2 ;
+const CURRENT_CONTRACT_DESCRIPTION_VERSION = 3 ;
```
Comment (freedomhero) [inline src/compilerWrapper.ts:20]: It seems that the desc file is not affected by this pr, so no need to upgrade version.

File: src/compilerWrapper.ts [hunk 2]
```diff
@@ -9,19 +9,18 @@ import { path2uri } from './utils';
 import compareVersions = require('compare-versions');
 
 const SYNTAX_ERR_REG = /(?<filePath>[^\s]+):(?<line>\d+):(?<column>\d+):\n([^\n]+\n){3}(unexpected (?<unexpected>[^\n]+)\nexpecting (?<expecting>[^\n]+)|(?<message>[^\n]+))/g;
-const SEMANTIC_ERR_REG = /Error:\s*(?<filePath>[^\s]+):(?<line>\d+):(?<column>\d+):(?<line1>\d+):(?<column1>\d+)\n(?<message>[^\n]+)\n/g;
-const IMPORT_ERR_REG_V1 = /Syntax error:\s*(?<filePath>[^\s]+):(?<line>\d+):(?<column>\d+):\n([^\n]+\n){3}File not found: (?<fileName>[^\s]+)/g;
-const IMPORT_ERR_REG_V2 = /Error:\s*\n(?<filePath>[^:]+):(?<startline>\d+):(?<startcolumn>\d+):(?<endline>\d+):(?<endcolumn>\d+):\nFile not found:\s*"(?<fileName>.+)"\n\n/g;
+const COMMON_ERR_REG = /Error:(\s|\n)*(?<filePath>[^\s]+):(?<line>\d+):(?<column>\d+):(?<line1>\d+):(?<column1>\d+):*\n(?<message>[^\n]+)\n/g;
```
Comment (freedomhero) [inline src/compilerWrapper.ts:12]: It's better called SEMANTIC_ERR

File: src/compilerWrapper.ts [hunk 3]
```diff
@@ -105,7 +99,7 @@ export enum ABIEntityType {
 	CONSTRUCTOR = 'constructor'
 }
 export type ParamEntity = {
-	name: string, type: string, finalType: string
+	name: string, type: string
```
Comment (freedomhero) [inline src/compilerWrapper.ts:102]: Follow others' format

File: src/compilerWrapper.ts [hunk 4]
```diff
@@ -221,11 +195,13 @@ export function compile(
 				return {
 					errors: syntaxErrors
 				};
-			} else {
-				const semanticErrors: CompileError[] = [...output.matchAll(SEMANTIC_ERR_REG)].map(match => {
+			} 
+			else {
+
+				const commonErrors: CompileError[] = [...output.matchAll(COMMON_ERR_REG)].map(match => {
```
Comment (freedomhero) [inline src/compilerWrapper.ts:201]: Same here, common -> semantic

File: src/utils.ts [hunk 5]
```diff
@@ -420,8 +420,13 @@ export function isStructType(type: string): boolean {
 	return /^struct\s(\w+)\s\{\}$/.test(type);
 }
 
-export function isArrayType(type: string) {
-	return /[^[\]]+\[\d+\]/.test(type);
+/**
+ * 
+ * @param type test struct Token {}[3], int[3], st.b.c[3]
```
Comment (freedomhero) [inline src/utils.ts:425]: Better not to use document comment format here, single line comment preferred.

File: src/contract.ts [hunk 6]
```diff
@@ -346,6 +348,7 @@ export function buildStructsClass(desc: CompileResult | ContractDescription): Re
       }
     });
 
+    structTypes[name].alias = desc.alias || [];
```
Comment (freedomhero) [inline src/contract.ts:351]: Duplicated alias on every struct class?

File: src/abi.ts [hunk 7]
```diff
@@ -177,7 +174,7 @@ export class FunctionCall {
 
 export class ABICoder {
 
-  constructor(public abi: ABIEntity[], public structs: StructEntity[]) { }
+  constructor(public abi: ABIEntity[], public alias: AliasEntity[]) { }
```
Comment (freedomhero) [inline src/abi.ts:177]: Why structs not needed any more? Feels like it should be there.

File: src/contract.ts [hunk 8]
```diff
@@ -358,38 +361,55 @@ export function buildTypeClasses(desc: CompileResult | ContractDescription): Rec
 
   const structClasses = buildStructsClass(desc);
```
Comment (freedomhero) [inline src/contract.ts:362]: Used deprecated method??

File: src/scryptTypes.ts [hunk 9]
```diff
@@ -9,15 +9,14 @@ export class ScryptType {
   protected _literal: string;
   protected _asm: string;
   protected _type: string;
-  protected _finalType: string;
+  public static alias: AliasEntity[] = [];
```
Comment (freedomhero) [inline src/scryptTypes.ts:12]: Better keep `_finalType` field bcoz it makes senson, `alias` should not be a member of ScryptType.

File: src/scryptTypes.ts [hunk 10]
```diff
@@ -29,7 +28,10 @@ export class ScryptType {
   }
 
   get finalType(): string {
-    return this._finalType;
+    const alias: AliasEntity[] =  Object.getPrototypeOf(this).constructor.alias;
```
Comment (freedomhero) [inline src/scryptTypes.ts:31]: Better do this in constructor function or outside of the class.

File: src/compilerWrapper.ts (hunk 1 above)
Comment (zhfnjust) [reply src/compilerWrapper.ts:20]: ast remove finalType affected  that desc file , no more finalType in desc

File: src/contract.ts (hunk 8 above)
Comment (zhfnjust) [reply src/contract.ts:362]: buildStructsClass is deprecated  , but it still work, buildTypeClasses is more powerful. we suggest users to using buildTypeClasses

File: src/scryptTypes.ts (hunk 9 above)
Comment (zhfnjust) [reply src/scryptTypes.ts:12]: refactor: using _typeResolver
18e4b4264b4ac49eb31218dad09ba0283c7cb83b

File: src/scryptTypes.ts (hunk 10 above)
Comment (zhfnjust) [reply src/scryptTypes.ts:31]: refactor: using _typeResolver
18e4b4264b4ac49eb31218dad09ba0283c7cb83b

**Pull request merged**